zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -219,44 +267,152 @@ where
+                    })
+                    .map(|(index, _)| index)
+                    .expect("An output should exist in the transaction for each transparent payment.")
+            }
+        };
+
+        SentTransactionOutput {
+            output_index: idx,
+            recipient_address: &payment.recipient_address,
+            value: payment.amount,
+            memo: payment.memo.clone()
         }
-    };
+    }).collect();
+
+    wallet_db.store_sent_tx(&SentTransaction {
+        tx: &tx,
+        created: time::OffsetDateTime::now_utc(),
+        outputs: sent_outputs,
+        account,
+        utxos_spent: vec![],
+    })
+}
+
+#[cfg(feature = "transparent-inputs")]
+#[allow(clippy::too_many_arguments)]
+pub fn shield_funds<E, N, P, D, R>(
+    wallet_db: &mut D,
+    params: &P,
+    prover: impl TxProver,
+    account: AccountId,
+    sk: &secp256k1::SecretKey,
+    extsk: &ExtendedSpendingKey,
+    memo: &MemoBytes,
+    confirmations: u32,
+) -> Result<D::TxRef, E>
+where
+    E: From<Error<N>>,
+    P: consensus::Parameters,
+    R: Copy + Debug,
+    D: WalletWrite<Error = E, TxRef = R>,
+{
+    let (latest_scanned_height, latest_anchor) = wallet_db
+        .get_target_and_anchor_heights()
+        .and_then(|x| x.ok_or_else(|| Error::ScanRequired.into()))?;
+
+    // derive the corresponding t-address
+    let taddr = derive_transparent_address_from_secret_key(sk);
+
+    // derive own shielded address from the provided extended spending key
+    let z_address = extsk.default_address().unwrap().1;
+
+    let exfvk = ExtendedFullViewingKey::from(extsk);
+
+    let ovk = exfvk.fvk.ovk;
+
+    // get UTXOs from DB
+    let utxos = wallet_db.get_unspent_transparent_utxos(&taddr, latest_anchor - confirmations)?;
+    let total_amount = utxos
+        .iter()
+        .map(|utxo| utxo.value)
+        .sum::<Option<Amount>>()
+        .ok_or_else(|| E::from(Error::InvalidAmount))?;
+
+    let fee = DEFAULT_FEE;
+    if fee >= total_amount {
+        return Err(E::from(Error::InsufficientBalance(total_amount, fee)));
+    }
+
+    let amount_to_shield = (total_amount - fee).ok_or_else(|| E::from(Error::InvalidAmount))?;
+
+    let mut builder = Builder::new(params.clone(), latest_scanned_height);
+
+    #[cfg(feature = "transparent-inputs")]
+    for utxo in &utxos {
+        let coin = TxOut {
+            value: utxo.value,
+            script_pubkey: Script {
+                0: utxo.script.clone(),
+            },
+        };
+
+        builder
+            .add_transparent_input(*sk, utxo.outpoint.clone(), coin)
+            .map_err(Error::Builder)?;
+    }
+
+    // there are no sapling notes so we set the change manually
+    builder.send_change_to(ovk, z_address.clone());
+
+    // add the sapling output to shield the funds
+    builder
+        .add_sapling_output(Some(ovk), z_address.clone(), amount_to_shield, memo.clone())
+        .map_err(Error::Builder)?;
```

Comment:
Hm. Removing the change address seems risky, because then if there's an error somewhere the rest gets sent to a miner. What we should actually do is have the transaction builder omit the change output if it's zero, if it doesn't already do that.